pkg/util/scheduler/scheduler.go

```diff
@@ -0,0 +1,203 @@
+	logger     log.Logger
+}
+
+func (w *Worker) run(ctx context.Context) {
+	defer w.wg.Done()
+	w.logger.Debug("worker started", "id", w.id)
+
+	for {
+		select {
+		case <-ctx.Done():
+			w.logger.Debug("worker stopped", "id", w.id)
+			return
+		default:
+			w.dequeueWithRetries(ctx)
+		}
+	}
+}
+
+func (w *Worker) dequeueWithRetries(ctx context.Context) {
+	boff := backoff.New(ctx, backoff.Config{
+		MinBackoff: DefaultMinBackoff,
+		MaxBackoff: w.maxBackoff,
+		MaxRetries: w.maxRetries,
+	})
+
+	for boff.Ongoing() {
+		runnable, err := w.queue.Dequeue(ctx)
+		if err == nil {
+			runnable()
+			break
+		}
+
+		if errors.Is(err, ErrQueueClosed) {
+			w.logger.Debug("queue closed, stopping worker", "id", w.id)
+			return
+		}
+
+		w.logger.Error("retrying dequeue", "id", w.id, "error", err, "attempt", boff.NumRetries())
+		boff.Wait()
+	}
+	if err := boff.ErrCause(); err != nil {
+		w.logger.Error("failed to dequeue after retries", "id", w.id, "error", err)
+		return
+	}
+}
+
+// Scheduler manages a pool of Workers consuming from a FairQueue.
+type Scheduler struct {
+	services.Service
+
+	logger     log.Logger
+	queue      WorkQueue
+	wg         sync.WaitGroup
+	maxBackoff time.Duration
+
+	subservices        *services.Manager
+	subservicesWatcher *services.FailureWatcher
+
+	workers    []*Worker
+	numWorkers int
+}
+
+// Config holds configuration for the Scheduler.
+type Config struct {
+	NumWorkers int
+	MaxBackoff time.Duration
+	MaxRetries int
+	Logger     log.Logger
+}
+
+func (c *Config) validate() error {
+	if c.NumWorkers <= 0 {
+		c.NumWorkers = DefaultNumWorkers
+	}
+	if c.MaxBackoff <= 0 {
+		c.MaxBackoff = DefaultMaxBackoff
+	}
+	if c.MaxRetries <= 0 {
+		c.MaxRetries = DefaultMaxRetries
+	}
+	if c.Logger == nil {
+		c.Logger = log.New("scheduler")
+	}
+	return nil
+}
+
+// NewScheduler creates a new scheduler instance.
+func NewScheduler(queue WorkQueue, config *Config) (*Scheduler, error) {
+	var err error
+
+	if queue == nil {
+		return nil, errors.New("queue cannot be nil")
+	}
+	if err := config.validate(); err != nil {
+		return nil, fmt.Errorf("invalid config: %w", err)
+	}
+
+	s := &Scheduler{
+		logger:             config.Logger,
+		queue:              queue,
+		numWorkers:         config.NumWorkers,
+		maxBackoff:         config.MaxBackoff,
+		workers:            make([]*Worker, 0, config.NumWorkers),
+		subservicesWatcher: services.NewFailureWatcher(),
+	}
+
+	subservices := []services.Service{s.queue}
+	s.subservices, err = services.NewManager(subservices...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create subservices manager: %w", err)
+	}
+	s.Service = services.NewBasicService(s.starting, s.running, s.stopping)
+	return s, nil
+}
+
+// starting is called by the services.Service lifecycle to start the scheduler.
+func (s *Scheduler) starting(ctx context.Context) error {
+	s.subservicesWatcher.WatchManager(s.subservices)
+
+	if err := services.StartManagerAndAwaitHealthy(ctx, s.subservices); err != nil {
```

Comment:
Since `queue` is passed as a dependency into `NewScheduler`, I don't think scheduler should be starting it. We could instead change `NewScheduler` to create the queue, if we want to start it here.

I would also perhaps drop subservices (since there's only one) and subservicesWatcher (because Queue cannot fail on its own, and will keep running until explicitly stopped).